subprojects/core/src/main/java/org/gradle/api/internal/project/taskfactory/TaskFactory.java

```diff
@@ -59,9 +61,21 @@ public ITaskFactory createChild(ProjectInternal project, InstantiationScheme ins
         NameValidator.validate(identity.name, "task name", "");
 
         final Class<? extends AbstractTask> implType;
-        if (identity.type.isAssignableFrom(DefaultTask.class)) {
+        if (identity.type == Task.class) {
+            implType = DefaultTask.class;
+        } else if (DefaultTask.class.isAssignableFrom(identity.type)) {
+            implType = identity.type.asSubclass(AbstractTask.class);
+        } else if (identity.type == AbstractTask.class || identity.type == TaskInternal.class) {
+            DeprecationLogger.deprecate("Registering a task with type " + identity.type.getSimpleName())
+                .willBecomeAnErrorInGradle7()
+                .withUpgradeGuideSection(6, "abstract_task_deprecated")
+                .nagUser();
             implType = DefaultTask.class;
         } else {
+            DeprecationLogger.deprecate("Registering a task with a subtype of AbstractTask")
```

Comment:
This is a bit imprecise, as `DefaultTask` is subtype of `AbstractTask`, too. How about something like:

```suggestion
            DeprecationLogger.deprecate("Registering a task that directly extends AbstractTask")
```